### PR TITLE
fix: fix time drift between client and server when redeeming v3

### PR DESCRIPTION
**Pull request by pavelbrm**
This PR adds a 1 hour leeway to the check for issuer v3 signing key, as occasionally clients might fail if requests come through around the time of key rotation.

It also fixes a bug in the v3 redemption handler where issuer expiry check would never work – it mistakenly required the expiry time be both zero and after now at the same time, which was clearly a mistake (as anywhere else that check requires the time to not be zero).

Additionally, the PR offers small refactors to improve readability and maintainability, as well as some test coverage.

For more details, please see the linked issue.

### CODE REVIEW

File: kafka/signed_token_redeem_handler.go [hunk 1]
```diff
@@ -91,10 +93,12 @@ func SignedTokenRedeemHandler(
 		return nil
 	}
 
-	// Create a lookup for issuers & signing keys based on public key
+	// Create a lookup for issuers & signing keys based on public key.
 	signedTokens := make(map[string]SignedIssuerToken)
+	now := time.Now()
+
 	for _, issuer := range issuers {
-		if !issuer.ExpiresAtTime().IsZero() && issuer.ExpiresAtTime().Before(time.Now()) {
```
Comment (pavelbrm) [inline kafka/signed_token_redeem_handler.go:97]: There is absolutely no need to:
- spread this piece of logic across multiple files;
- calculate expiry time two times each issuer;
- calculate `now` on each iteration.

File: btd/issuer.go [hunk 2]
```diff
@@ -113,32 +113,39 @@ func ApproveTokens(blindedTokens []*crypto.BlindedToken, key *crypto.SigningKey)
...
+
 			return err
 		}
+
+		_ = timerVrf.ObserveDuration()
+
 		if valid {
 			break
 		}
-		timer.ObserveDuration()
```
Comment (pavelbrm) [inline btd/issuer.go:141]: This timer wasn't quite triggered upon errors or successful validation. 🤦‍♂️

**Comment by pavelbrm** [comment]
> [[puLL-Merge](https://github.com/brave/pull-merge)] - [brave-intl/challenge-bypass-server@711](https://github.com/brave-intl/challenge-bypass-server/pull/711)
> 

Again robot, you are disappointing. 🤦‍♂️ 

> 1. `btd/issuer.go` - The `VerifyTokenRedemption` function now derives the unblinded token and shared key separately for each potential signing key. This reduces the risk of timing attacks that could reveal information about which key was used. However, the separate timing measurements should be carefully reviewed to ensure they do not introduce any new timing side channels.

This has existed, and has not been changed other than a few comments and fixing a couple of issues. You must learn and tell other developers that re-using the same variable for different values is bad. And you should also teach them that observing time on timers needs to happen also on successful and error cases.

> 2. `model/issuer.go`, `model/issuer_keys.go` - The new functionality for finding active signing keys based on their validity time window looks solid. The tests help ensure correctness. One suggestion would be to add some logging if no active keys are found for an issuer, as that could indicate a configuration issue or key rotation problem.

There are already way too many semi-competent engineers who stick logging mostly in inappropriate places. And with your guidance, we'll get even worse software, and soon – more of such engineers. Please go learn about functions without side effects, and learn to categorise logging as a side effect.

File: model/issuer_test.go [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+			if tc.exp.err != nil {
+				return
+			}
+
+			should.Equal(t, tc.exp.result, actual)
+		})
+	}
+}
+
+func TestIssuer_findActiveKeysEV(t *testing.T) {
```
Comment (pavelbrm) [inline model/issuer_test.go:303]: @evq Here's a test that makes sure the conditions we've discussed are met.

Comment (pavelbrm) [reply model/issuer_test.go]: Disregard this comment.

File: model/issuer_test.go [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+					{
+						StartAt: ptrTo(time.Date(2024, time.January, 1, 0, 0, 1, 0, time.UTC)),
+						EndAt:   ptrTo(time.Date(2024, time.January, 2, 0, 0, 1, 0, time.UTC)),
+					},
+				},
+			},
+		},
+
+		{
+			name: "evq_strict_b_leeway_a",
```
Comment (pavelbrm) [inline model/issuer_test.go:287]: @evq This test case and the case below it are the two we discussed last week.

**Review by Sneagan — APPROVED** [review]

**Review by evq — APPROVED** [review]

**Pull request merged**